packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,208 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import {
+	ISequencedDocumentMessage,
+	ISnapshotTree,
+	IVersion,
+} from "@fluidframework/protocol-definitions";
+import { IGetPendingLocalStateProps, IRuntime } from "@fluidframework/container-definitions";
+import {
+	ITelemetryLoggerExt,
+	MonitoringContext,
+	PerformanceEvent,
+	UsageError,
+	createChildMonitoringContext,
+} from "@fluidframework/telemetry-utils";
+import { assert } from "@fluidframework/core-utils";
+import { IResolvedUrl, ISnapshot } from "@fluidframework/driver-definitions";
+import { isInstanceOfISnapshot } from "@fluidframework/driver-utils";
+import {
+	ContainerStorageAdapter,
+	ISerializableBlobContents,
+	getBlobContentsFromTree,
+} from "./containerStorageAdapter";
+import { IPendingContainerState } from "./container";
+
+export class containerStateManager {
+	private readonly savedOps: ISequencedDocumentMessage[] = [];
+	public snapshot:
+		| {
+				tree: ISnapshotTree;
+				blobs: ISerializableBlobContents;
+		  }
+		| undefined;
+	private readonly mc: MonitoringContext;
+	private readonly clientId: string | undefined;
```

Comment:
clientId can change, so i don't know what a immutable id provides. i would expect something like `getClientId: ()=>string` that returns the current clientId of the object that owns this object, which in prod will be the container, but in test can be anything